web/src/app/components/bar-chart/bar-chart.component.ts

```diff
@@ -0,0 +1,357 @@
+import { AbstractType, AfterViewInit, Component, ElementRef, Input, OnChanges, SimpleChanges, ViewChild} from '@angular/core';
+import { formatDate } from '@angular/common'
+import * as d3 from 'd3';
+
+import { step189_2020 } from '../../../proto/step189_2020';
+
+/**
+ * Item includes all data used by the bar chart.
+ */
+interface Item {
+  pushID: string;    // Push ID number
+  state: number;     // Tag of push end state
+  startTime: string; // Start time of push, in `YYYY-MM-DD HH:mm` format
+  duration: number;  // Duration of push, in hour
+}
+
+/**
+ * Define D3 type used by the bar chart.
+ */
+type d3SVG = d3.Selection<SVGSVGElement, undefined, null, undefined>;
+type d3G = d3.Selection<SVGGElement, undefined, null, undefined>;
+type d3ScaleLinear = d3.ScaleLinear<number, number>;
+
+@Component({
+  selector: 'app-bar-chart',
+  templateUrl: './bar-chart.component.html',
+  styleUrls: ['./bar-chart.component.scss']
+})
+
+export class BarChartComponent implements AfterViewInit {
+  /**
+   * Constants.
+   */
+  private static readonly DEFAULT_NUM_BARS: number = 30;
+  private static readonly NANO_TO_SECS: number = 10 ** 9;
+  private static readonly NANO_TO_MILLI: number = 10 ** 6;
+  private static readonly SECS_TO_HRS: number = 60 * 60;
+  private static readonly COLOR_LIGHT_GRAY: string = '#787878';
+  private static readonly COLOR_DARK_GRAY: string = '#373C38';
+  private static readonly COLOR_WHITE: string = '#eee';
+  private static readonly COLOR_WHITE_TRANS: string = '#ffffff00';
+  private static readonly DATE_FORMAT: string = 'yyyy-MM-dd HH:mm';
```

Comment:
'yyyy-MM-dd HH:mm' Should the letters be capitalized or not? Prefer to make them consistent.